File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
           ListNode prev = null;
        ListNode curr = head;
        
        while (curr != null) {
            ListNode nextNode = curr.next;  // store next node
            curr.next = prev;               // reverse the link
            prev = curr;                    // move prev to current
            curr = nextNode;                // move current to next
        }
        
        return prev;  // new head of reversed list
    }
}
    

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {

        unordered_map<Node*, Node*> m;

        if (head == NULL) return NULL;
        
        Node *newHead = new Node(head->val);
        Node *oldTemp= head->next;
        Node *newTemp = newHead;
        m[head] = newHead;

        while(oldTemp!= NULL){
            Node *copynode = new Node(oldTemp->val);
            m[oldTemp] = copynode;
            newTemp->next = copynode;
            oldTemp= oldTemp->next;
            newTemp= newTemp->next;
        }

        oldTemp = head; newTemp = newHead;

        while(oldTemp!= NULL){
            newTemp->random = m[oldTemp ->random];
            oldTemp= oldTemp->next;
            newTemp= newTemp->next;


        }
        return newHead;
    }
};
